src/Hedgehog/Plutus/Tx.hs

```diff
@@ -0,0 +1,600 @@
+    case spend of
+      Nothing -> pure detSpend
+      -- if the shuffled balance somehow fails
+      -- just return the deterministic spend
+      Just s -> pure s
+  where
+    nonSpecial :: Plutus.TxOut -> Bool
+    nonSpecial txout =
+      let
+        cred = Plutus.addressCredential $ Plutus.txOutAddress txout
+        noDatum = Plutus.txOutDatum txout == Plutus.NoOutputDatum
+        isPubKey = case cred of
+          Plutus.PubKeyCredential _ -> True
+          _ -> False
+       in
+        isPubKey && noDatum
+
+    toRes :: (Value, Spend) -> Maybe Spend
+    toRes (rem, s) = guard (rem == mempty) $> s
+
+    go ::
+      (Value, Spend) ->
+      (Plutus.TxOutRef, Plutus.TxOut) ->
+      (Value, Spend)
+    go (remaining, s@(Spend refs change)) (outRef, out)
+      | remaining == mempty = (remaining, s)
+      | intersect /= mempty =
+          (remaining', Spend (Vector.cons outRef refs) change')
+      | otherwise = (remaining, s)
+      where
+        refVal = Plutus.txOutValue out
+        intersect = Plutus.unionWith min remaining refVal
+        remaining' = posDif remaining intersect
+        change' = change <> posDif refVal intersect
+
+{- | Generate a @plutus-simple-model@ 'Model.Tx' from a @hedgehog-plutus-simple@
+ 'Tx'. This should automatically add neccesary scripts and signatures.
+-}
+toSimpleModelTx :: TxContext -> Tx bal -> Model.Tx
+toSimpleModelTx
+  TxContext
+    { interestingScripts
+    , mockchain
+    , datums
+    }
+  Tx
+    { txInputs
+    , txOutputs
+    , txMint
+    , txFee
+    , txValidRange
+    , txExtraSignatures
+    } =
+    Model.Tx
+      { Model.tx'extra = mempty
+      , Model.tx'plutus =
+          mempty
+            { Fork.txInputs = Set.map convertTxIn txInputs
+            , Fork.txCollateral = mempty
+            , -- TODO our tx type should probably support reference inputs
+              Fork.txReferenceInputs = mempty
+            , Fork.txOutputs = outs
+            , Fork.txCollateralReturn = Nothing
+            , Fork.txTotalCollateral = Nothing
+            , Fork.txMint =
+                mconcat
+                  [ singleton cs tn amt
+                  | (cs, (toks, _reds)) <- Map.toList txMint
+                  , (tn, amt) <- Map.toList toks
+                  ]
+            , Fork.txFee = txFee
+            , Fork.txValidRange =
+                Time.posixTimeRangeToContainedSlotRange
+                  (Model.mockConfigSlotConfig $ Model.mockConfig mockchain)
+                  txValidRange
+            , Fork.txMintScripts =
+                Set.fromList $
+                  [ scriptToMP $ getScript $ Plutus.ScriptHash cs
+                  | (Plutus.CurrencySymbol cs, _) <- Map.toList txMint
+                  ]
+            , Fork.txSignatures =
+                Map.fromList
+                  [ (pkh, Model.userSignKey $ getUser pkh)
+                  | pkh <- Set.toList txExtraSignatures
+                  -- TODO if these are "extra" are the non-extra just
```

Comment:
Part of the job of this layer is to resolve all pubkeyhashes from inputs. the extraSignatures field is only for any additional signatures a script may require.

Could you add a haddock to the field explaining this?